samples/nrf9160/modem_shell/src/th/th_ctrl.c

```diff
@@ -0,0 +1,316 @@
+		shell_print(shell, "  Results available");
+		if (k_work_is_pending(&(data->work))) {
+			shell_print(shell, "  thread is running");
+		} else {
+			shell_print(shell, "  thread is not running");
+		}
+		print_buf = (char *)calloc(data->cmd_len + 1, sizeof(char));
+		shell_print(shell, "  command: %s",
+			    th_ctrl_get_command_str_from_argv(
+				    data->argc, data->argv, print_buf,
+				    data->cmd_len + 1));
+		free(print_buf);
+	} else {
+		shell_print(shell, "  Nothing");
+	}
+}
+
+void th_ctrl_status_print(const struct shell *shell)
+{
+	th_ctrl_data_status_print(shell, &th_work_data_1);
+	th_ctrl_data_status_print(shell, &th_work_data_2);
+}
+
+void th_ctrl_kill(const struct shell *shell, int nbr)
+{
+	if (nbr == 1) {
+		if (k_work_is_pending(&(th_work_data_1.work))) {
+			k_poll_signal_raise(&th_work_data_1.kill_signal, 1);
+		} else {
+			shell_print(shell, "Thread #1 not running");
+			return;
+		}
+	} else if (nbr == 2) {
+		if (k_work_is_pending(&(th_work_data_2.work))) {
+			k_poll_signal_raise(&th_work_data_2.kill_signal, 2);
+		} else {
+			shell_print(shell, "Thread #2 not running");
+			return;
+		}
+	}
+}
+
+static void th_ctrl_data_result_print(const struct shell *shell,
+				      struct th_ctrl_data *data)
+{
+	if (data->results_str == NULL || !strlen(data->results_str)) {
+		shell_print(shell, "No results for thread #%d", data->th_nbr);
+	} else {
+		shell_print(shell, "thread #%d results:", data->th_nbr);
+		shell_print(shell, "-------------------------------------");
+		shell_print(shell, "%s", data->results_str);
+		shell_print(shell, "-------------------------------------");
+
+		/* Delete data if the work is done: */
+		if (!k_work_is_pending(&(data->work))) {
+			free(data->results_str);
+			data->results_str = NULL;
+			shell_print(shell, "Note: th results #%d were deleted.",
+				    data->th_nbr);
+
+			/* Clean up for cmd argv: */
+			th_ctrl_util_duplicate_argv_free(data->argc,
+							 data->argv);
+			data->argc = 0;
+			data->argv = NULL;
+		}
+	}
+}
+
+void th_ctrl_result_print(const struct shell *shell, int nbr)
+{
+	if (nbr == 1) {
+		th_ctrl_data_result_print(shell, &th_work_data_1);
+	} else if (nbr == 2) {
+		th_ctrl_data_result_print(shell, &th_work_data_2);
+	}
+}
+
+static void th_ctrl_data_start(struct th_ctrl_data *data,
+			       const struct shell *shell, size_t argc,
+			       char **argv, bool is_background)
+{
+	assert(!k_work_is_pending(&(data->work)));
+
+	if (!is_background && data->results_str != NULL) {
+		/* Flush and free result "file for foreground: */
+		free(data->results_str);
+		data->results_str = NULL;
+	}
+
+	if (is_background && data->results_str == NULL) {
+		data->results_str =
+			(char *)calloc(TH_RESPONSE_BUFFER_SIZE, sizeof(char));
+		if (data->results_str == NULL) {
+			shell_error(
+				shell,
+				"Cannot start a thread: no memory to store a response");
+			return;
+		}
+	}
+
+	if (data->results_str != NULL) {
```

Comment:
This could probably be an "else if" for the above "if", so that the buffer allocated using `calloc()` wouldn't be again zeroed using `memset()`.

Reply:
Thanks for all comments. All fixed now.